alf/environments/metadrive/extra_rewards.py

```diff
@@ -0,0 +1,348 @@
+        self._history_speed = np.zeros(3)
+        self._history_heading = np.zeros(5)
+
+    def env_info_spec(self):
+        return {
+            "MetaDrive/harsh_brake":
+                TensorSpec(shape=(), dtype=torch.float32),
+            "MetaDrive/lon_acc":
+                TensorSpec(shape=(), dtype=torch.float32),
+            "MetaDrive/lon_jerk":
+                TensorSpec(shape=(), dtype=torch.float32),
+            "MetaDrive/lat_acc":
+                TensorSpec(shape=(), dtype=torch.float32),
+            "MetaDrive/lat_jerk":
+                TensorSpec(shape=(), dtype=torch.float32),
+            "MetaDrive/costs/lon_acc":
+                TensorSpec(shape=(), dtype=torch.float32),
+            "MetaDrive/costs/lon_jerk":
+                TensorSpec(shape=(), dtype=torch.float32),
+            "MetaDrive/costs/lat_jerk":
+                TensorSpec(shape=(), dtype=torch.float32),
+        }
+
+
+@alf.configurable
+class LaneKeepingReward(ExtraReward):
+    """The reward that penalizes riding a broken line.
+
+    A broken line resides between two lanes that permits lane change. This
+    reward particularly encourages the car to stay in the lane unless it wants
+    to perform lane change or overtaking.
+
+    Args:
+
+        broken_line_cost: the penalty for one step if the ego car is riding on
+            a broken line.
+
+    """
+
+    def __init__(self, broken_line_cost: float = 0.05):
+        super().__init__()
+        self._broken_line_cost = broken_line_cost
+
+    def evaluate(self, engine: BaseEngine):
+        ego: BaseVehicle = engine.managers["agent_manager"].active_agents[
+            "default_agent"]
+
+        rewards = {
+            "on_broken_line":
+                -self._broken_line_cost if ego.on_broken_line else 0.0
+        }
+
+        info = {"MetaDrive/on_broken_line": float(ego.on_broken_line)}
+
+        return rewards, info
+
+    def env_info_spec(self):
+        return {
+            "MetaDrive/on_broken_line":
+                TensorSpec(shape=(), dtype=torch.float32)
+        }
+
+
+@alf.configurable
+class CrashVehicleReward(ExtraReward):
+    """The EXTRA reward that penalizes ego car from crashing into another vehicle.
+
+    MetaDrive has already has a flat penalty towards all kinds of crashing (road boundary,
+    vehicle, objects, etc). To further penalize crashing into vehicles over crashing into
+    the road boundary, we added this extra reward.
+
+    Note that the episode will END when a crash happens on the ego car. This means that
+    such reward is imposed at most ONCE per episode.
+
+    Args:
+
+        cost: the extra cost imposed when the crash is with another vehicle.
+
+    """
+
+    def __init__(self, cost: float = 20.0):
+        super().__init__()
+        self._cost = cost
+
+    def evaluate(self, engine: BaseEngine):
+        ego: BaseVehicle = engine.managers["agent_manager"].active_agents[
+            "default_agent"]
+        cost = 0.0
```

Comment:
nitpick: Since this reward is very similar to LaneKeepingReward, I feel the code (line 336-342 and. line 296-303) should also be written in a similar way.

Reply:
Very reasonable. Updated to make them consistent.